### PR TITLE
Add GameDig monitor

### PR DESCRIPTION
**Pull request by WhyKickAmooCow**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]: 
- [x] I have read and understand the pull request rules.

# Description
This pull request adds support for using GameDig as a monitor, this allows rich game server monitoring rather than the apparently unreliable method of using the TCP monitor.

Fixes #(issue)
#1053
#479
#525

## Type of change

- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [ ] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

Please do not use any external image service. Instead, just paste in or drag and drop the image here, and it will be uploaded automatically.

### CODE REVIEW

**Comment by louislam** [comment]
https://github.com/gamedig/node-gamedig/blob/master/games.txt

This monitor will be perfect if the game field is a dropdown instead of a free text field.

**Comment by WhyKickAmooCow** [comment]
Autofill in the port may not be necessary. When you tell gamedig the game type, it will automatically resolve the port to use (defaults and through SRV records).

A game dropdown would be great though, I guess the dropdown list would have to be an object with an ID and a displayname so it is easy to select a game by name without knowing the ID.

**Comment by cetteup** [comment]
A gamedig monitor would we **great**.

@WhyKickAmooCow Are you working on the game dropdown?

gamedig's `games.txt` already contains an id and a human readable name, the id is what you'd pass to the query as `type`.

Regarding the port: I'd say letting gamedig guess the port every single time is not ideal. Especially because gamedig uses a port offset for some games, which it always applies (see [lib/QueryRunner.js#L33](https://github.com/gamedig/node-gamedig/blob/57a6a2fadd2048afd9c38ee2b2b095193e4039ae/lib/QueryRunner.js#L33)). For a Battlefield 4 server using query port 47200, gamedig
- attempts to query the server on given port (47200) + query port offset (22000)
- attempts to query server on given port (47200)
- attempts to query server on default game port (25200) + query port offset

Since server admins generally know their query port settings (or can find out rather easily). Thus I would say a much better option is to fill the port field with a default (which can be determined based on data from `games.txt`) and use the `givenPortOnly` option for the query. If the port is wrong, the monitor runs a single query that fails. If the port is correct (and the game server is up), the monitor runs a single query that succeeds.

**Numer of queries sent (Battlefield 4 example)**
`givenPortOnly`|Server up|Server down
---------------------|-------------|----------------
`false` (default)|2|3
`true`|1|1

**Comment by louislam** [comment]
> A gamedig monitor would we **great**.
> 
> @WhyKickAmooCow Are you working on the game dropdown?
> 
> gamedig's `games.txt` already contains an id and a human readable name, the id is what you'd pass to the query as `type`.
> 
> Regarding the port: I'd say letting gamedig guess the port every single time is not ideal. Especially because gamedig uses a port offset for some games, which it always applies (see [lib/QueryRunner.js#L33](https://github.com/gamedig/node-gamedig/blob/57a6a2fadd2048afd9c38ee2b2b095193e4039ae/lib/QueryRunner.js#L33)). For a Battlefield 4 server using query port 47200, gamedig
> 
>     * attempts to query the server on given port (47200) + query port offset (22000)
> 
>     * attempts to query server on given port (47200)
> 
>     * attempts to query server on default game port (25200) + query port offset
> 
> 
> Server admins generally know their query port settings (or can find out rather easily). Thus I would say a much better option is to fill the port field with a default (which can be determined based on data from `games.txt`) and use the `givenPortOnly` option for the query. If the port is wrong, the monitor runs a single query that fails. If the port is correct (and the game server is up), the monitor runs a single query that succeeds.
> 
> **Numer of queries sent (Battlefield 4 example)**
> `givenPortOnly` 	Server up 	Server down
> `false` (default) 	2 	3
> `true` 	1 	1

Thanks, maybe I try to add today.

**Comment by louislam** [comment]
Nice, it is more easy to use with the dropdown now.

image

File: server/model/monitor.js [hunk 1]
```diff
@@ -488,6 +490,20 @@ class Monitor extends BeanModel {
                     } else {
                         throw new Error("Server not found on Steam");
                     }
+                } else if (this.type === "gamedig") {
+                    try {
+                        const state = await Gamedig.query({
+                            type: this.game,
+                            host: this.hostname,
+                            port: this.port
+                        });
```
Comment (cetteup) [inline server/model/monitor.js:499]: If you add `givenPortOnly: true` here, you can avoid the multiple query issue I described.

**Comment by cetteup** [comment]
Really nice! Do you plan on adding the `givenPortOnly` option to the query? I'll gladly open a PR for that if need be, but it seems a bit overkill for one line of code.

**Comment by louislam** [comment]
Oh I forgot to add this. Should be ok now: https://github.com/louislam/uptime-kuma/commit/9cc3bd0de416bb12dcb1b156953b203b2fcd3fa8

**Comment by dgibbs64** [comment]
Just seen this has been added. Great news. Thank you

**Pull request merged**